sourcegraph/cody
# feature: custom prompts (internal)

RE: https://sourcegraph.slack.com/archives/C05869U2E7Q/p1686351275606069
More context: https://sourcegraph.slack.com/archives/C05AQ5LPAH1/p1689063906063269

## Summary

**Important - this feature is only available to s2 instance users for dogfooding purpose**

This PR adds a new feature that allows users to add custom recipes to Cody via:
- `.vscode/cody.json`: a json file from their current workspace, which can be shared with others easily (workspace recipes)
- `+` button that allows users to create new recipe (without commands) via UI (user recipes)
![image](https://github.com/sourcegraph/cody/assets/68532117/86d716ce-312d-4750-bac3-3d163587c11d)

This also makes testing prompts and preambles easier for our devs.

- lib/shared/src/chat/recipes/my-prompt.ts 
  - Recipe to run the selected custom prompt. It gets the prompt based on the custom recipe ID from the MyPromptController. If output of a command is available, it will also add the output as one of the context
- vscode/src/my-cody/MyPromptController.ts
  - The controller manages all the custom prompts, from retrieving the prompts from JSON file when the extension is first initialized to sharing the prompts with the `my-prompt` recipe
- vscode/src/my-cody/MyToolsProvider.ts
  - A very basic implementation that allows each custom recipe to use a `tool`, which for now is used for running commands defined in the JSON file

Video Demo: [Loom](https://www.loom.com/share/a0c11d79f1614611bfe7de72529b6922)

## Test Plan

- Launch Cody from this branch
- Connect Cody to S2 instance (Only available to s2 instance users for dogfooding)
- Create a cody.json file in the .vscode directory located in the root of your repository
- Add the example context below to the new `.vscode/cody.json\ file you just created, save the file
- Go to the recipes tab, you should be able to see the custom recipes from the cody.json file
- Highlight code in editor and then select a custom recipes
- Click on the + button to add a new recipe to global storage
- Switch to a non-S2 instance, the recipes should now disappeared

#### Example cody.json file

```json
{
  "recipes": {
    "spell checker": {
      "prompt": "Spell check my code and let me know if I have any typos"
    },
    "last git commit": {
      "prompt": "Who made the last commit to this repository?",
      "command": "git",
      "args": ["log", "-1"]
    },
    "code improve": {
      "prompt": "What changes should I make to improve the code?"
    },
    "GH Search: Deploy Repoes not owned by Sourcegraph": {
      "prompt": "How many repositories are there that named sourcegraph with the word deploy that is not owned by Sourcegraph?",
      "command": "gh",
      "args": ["search", "repos", "sourcegraph", "deploy"],
      "context": {
        "codebase": false
      }
    }
  }
}

```

## Follow up

This feature is currently in dogfood stage and unavailable to external users.

- [ ] button to remove custom recipe from UI
- [ ] tooltip on hover https://github.com/sourcegraph/cody/issues/243
- [x] Create recipes that support commands via `+` button
- [ ] instructions on creating workspace recipe and user recipe
